ethcore/src/client/client.rs

```diff
@@ -1450,26 +1454,24 @@ impl ImportBlock for Client {
 		}
 
 		let raw = if self.importer.block_queue.is_empty() {
-			Some((
-				unverified.bytes.clone(),
-				unverified.header.hash(),
-				*unverified.header.difficulty(),
-			))
-		} else { None };
+			Some((unverified.bytes.clone(), *unverified.header.difficulty()))
+		} else {
+			None
+		};
 
 		match self.importer.block_queue.import(unverified) {
 			Ok(hash) => {
-				if let Some((raw, hash, difficulty)) = raw {
-					self.notify(move |n| n.block_pre_import(&raw, &hash, &difficulty));
+				if let Some((bytes, difficulty)) = raw {
+					self.notify(move |n| n.block_pre_import(&bytes, &hash, &difficulty));
 				}
 				Ok(hash)
 			},
 			// we only care about block errors (not import errors)
-			Err((block, EthcoreError::Block(err))) => {
-				self.importer.bad_blocks.report(block.bytes, format!("{:?}", err));
-				return Err(EthcoreError::Block(err))
+			Err((EthcoreError::Block(err), input)) => {
+				self.importer.bad_blocks.report(input.map_or(Bytes::new(), |i| i.bytes), err.to_string());
```

Comment:
will it report this https://github.com/paritytech/parity-ethereum/blob/e1d06efe60afdd793cb6c9b599fbc9e8c76ac7c1/ethcore/src/client/bad_blocks.rs#L64

if the input is `None`? Can `input` be `None`? It's not supposed to be, right?
Do you think it makes sense to use `expect` here instead?

Reply:
Currently, the only condition when `input is None` is for the introduced `ImportError::AlreadyQueued` so I think it could be an expect.

However, it might be harder to reason about when maintaining/refactoring the code i.e, if a `(BlockError, None)` is introduced and then the node would panic just because the `raw bytes` of the error was missing (but it would still be a bug but not that severe as panic). How about a warning message? 

Ideally, this would be encoded in type-system (as you suggested) and we wouldn't need to worry about this.

Reply:
I would just use an `expect` for now and maybe add some docs to the `import` method (that for `EthcoreError::Block` we *should* return `Some(input)` as well as add a TODO here referencing an issue (and file an issue) to encode this in the type system.

Reply:
Sounds reasonable to me